src/engine/Match_search_mode.ml

```diff
@@ -702,186 +702,201 @@ and get_nested_formula_matches env formula range =
-      in
-      let all_posr =
-        match posrs @ posrs_inside with
-        | [] -> (
-            match opt_context with
-            | None -> failwith "empty And; no positive terms in And"
-            | Some r -> [ [ r ] ])
-        | ps -> ps
-      in
-      match all_posr with
-      | [] -> failwith "FIXME"
-      | posr :: posrs ->
-          let ranges = posr in
-          let ranges =
-            posrs
-            |> List.fold_left
-                 (fun acc r ->
-                   RM.intersect_ranges env.xconf.config
-                     ~debug_matches:!debug_matches acc r)
-                 ranges
-          in
-          (* optimization of `pattern: $X` *)
-          let ranges = run_selector_on_ranges env selector_opt ranges in
-
-          (* let's remove the negative ranges *)
-          let ranges, negs_expls =
-            neg
-            |> List.fold_left
-                 (fun (ranges, acc_expls) (tok, x) ->
-                   let ranges_neg, expl = evaluate_formula env opt_context x in
-                   let ranges =
-                     RM.difference_ranges env.xconf.config ranges ranges_neg
-                   in
-                   let expl =
-                     if_explanations env ranges [ expl ] (OutJ.Negation, tok)
-                   in
-                   (ranges, expl :: acc_expls))
-                 (ranges, [])
-          in
-          (* let's apply additional filters.
-           * TODO: Note that some metavariable-regexp may be part of an
-           * AND where not all patterns define the metavar, e.g.,
-            *   pattern-inside: def $FUNC() ...
-           *   pattern: return $X
-           *   metavariable-regexp: $FUNC regex: (foo|bar)
-           * in which case the order in which we do the operation matters
-           * (at this point intersect_range will have filtered the
-           *  range of the pattern_inside).
-           * alternative solutions?
-           *  - bind closer metavariable-regexp with the relevant pattern
-           *  - propagate metavariables when intersecting ranges
-           *  - distribute filter_range in intersect_range?
-           * See https://github.com/returntocorp/semgrep/issues/2664
-           *)
-          let ranges, filter_expls =
-            conds
-            |> List.fold_left
-                 (fun (ranges_with_bindings, acc_expls) (tok, cond) ->
-                   let ranges_with_bindings =
-                     filter_ranges env ranges_with_bindings cond
-                   in
-                   let expl =
-                     if_explanations env
-                       (List_.map fst ranges_with_bindings)
-                       []
-                       (OutJ.Filter (Tok.content_of_tok tok), tok)
-                   in
-                   (ranges_with_bindings, expl :: acc_expls))
-                 (List_.map (fun x -> (x, [])) ranges, [])
-          in
+  let rec aux env opt_context ({ f; focus; conditions } : Rule.formula) =
```

Comment:
I would try to avoid recursive functions inside recursive functions, and find a better name than `aux`.

Reply:
I may be missing something but `evaluate_formula` would become just a `let rec aux ... and aux' ... in aux ...`, so IMO definitely better to give proper names to `aux` and `aux'` and lift them one level up avoiding nested `let-rec`.